mira/sources/system_dynamics/pysd.py

```diff
@@ -280,91 +287,68 @@ def template_model_from_pysd_model(
     # template model is of type ControlledConversion for the SIR model; however, using sets,
     # the first type of template is of type NaturalConversion. Would require a lot of rewriting
     # of tests.
-    rates = []
+    rates = set()
     for state_rates in state_rate_map.values():
         for input_rates in state_rates["input_rates"]:
-            rates.append(input_rates)
+            rates.add(input_rates)
         for output_rates in state_rates["output_rates"]:
-            rates.append(output_rates)
+            rates.add(output_rates)
 
     # create map of transitions
-    for rate_name in rates:
-        # because we use a list of rates to not change the template order for testing purposes
-        # skip adding previously parsed rates to the transition_map
-        if rate_name in transition_map:
-            continue
-        inputs, outputs, controllers, rate_expr = [], [], [], None
-        for state_name, in_out_rate_map in state_rate_map.items():
-            # if a rate cannot be parsed, assign None
-            # This is a Stella issue as some expressions aren't created properly for rates
-            try:
-                rate_expr = safe_parse_expr(
-                    processed_expression_map[rate_name],
-                    symbols,
-                )
-            except TypeError:
-                rate_expr = SYMPY_FLOW_RATE_PLACEHOLDER
-
+    for rate_id in sorted(rates):
+        rate_expr = identifier_to_expr[rate_id]
+        inputs, outputs, controllers = [], [], []
+        for state_id, in_out_rate_map in state_rate_map.items():
             # if a rate is leaving a state, then that state is an input to the rate
-            if rate_name in in_out_rate_map["output_rates"]:
-                inputs.append(state_name)
+            if rate_id in in_out_rate_map["output_rates"]:
+                inputs.append(state_id)
 
             # if a rate is going into a state, then that state is an output to the rate
-            if rate_name in in_out_rate_map["input_rates"]:
-                outputs.append(state_name)
+            if rate_id in in_out_rate_map["input_rates"]:
+                outputs.append(state_id)
 
+                # FIXME this never happens
                 # if a state is present in a rate law, and the state isn't an input to the rate
                 # law, then that state is a controller of the rate law
-                if (
```

Comment:
@nanglo123 please resurrect this, and maybe we can add some unit tests that correspond to when this gets used